### PR TITLE
WpfPlot: Show plot in design mode

### PR DESCRIPTION
**Pull request by swharden**
addresses #1043

### CODE REVIEW

**Comment by swharden** [comment]
FormsPlot and WpfPlot now both show plots at design time.

If there is a problem with `System.Drawing.Common`, it may crash Visual Studio ☠️ 

https://user-images.githubusercontent.com/4165489/119075793-650a5480-b9bf-11eb-85e1-65515e309f31.mp4

**Comment by swharden** [comment]
I added a render test inside a try/catch block that should identify render errors and display them as text in the control at design time
image

**Comment by swharden** [comment]
WpfPlot and FormsPlot now do this as the first thing in their initializers:

```cs
string renderErrorMessage = ScottPlot.Drawing.GDI.DrawingTest();
if (renderErrorMessage != null)
{
    InitializeComponent();
    /* display the error message in a textbox */
    return;
}
```

@bclehmann do you think this is worth adding to AvaPlot?

File: src/ScottPlot/Drawing/GDI.cs [hunk 1]
```diff
@@ -18,6 +18,34 @@ public static class GDI
...
+                using var gfx = GDI.Graphics(bmp);
+                gfx.Clear(Color.Magenta);
+                gfx.DrawLine(Pens.Black, 0, 0, 10, 10);
+
+                using var stream = new System.IO.MemoryStream();
+                bmp.Save(stream, ImageFormat.Bmp);
+                byte[] bytes = stream.ToArray();
+                int expectedBitmapSize = 454;
+                if (bytes.Length != expectedBitmapSize)
+                    throw new InvalidOperationException($"test image was {bytes.Length} bytes (expected {expectedBitmapSize})");
```
Comment (bclehmann) [inline src/ScottPlot/Drawing/GDI.cs:39]: Throwing something in order to immediately catch it is something of an antipattern. It's using exception handling as flow control instead of an if/else.

This method should probably either return a boolean to indicate success/failure, or just let any exceptions bubble up to the caller.

If the actual error message is important then it should throw an exception which is caught by the caller and the error message surfaced to the user. Alternatively, you can return false to indicate failure and then set an error field somewhere (like in many C APIs), or return a struct with a success field set to false and some extra information (like in many AJAX libraries).

I don't normally like the C way of doing things because it's used everywhere to get around the lack of exception support, however it does have some benefits:
- It does not obligate the caller to handle any exceptions
    - Thus it's a pattern often used in AJAX when it doesn't matter if the request fails (or when failures are already handled by some middleware which will retry)
- Performance and compatibility with embedded systems (not really a consideration here)

Comment (bclehmann) [reply src/ScottPlot/Drawing/GDI.cs:39]: Returning a struct like this would probably be the best option, provided you don't want to let the exception bubble up to the caller. That choice depends on whether you think the caller ought to be obligated to handle the issue or die.

```cs
struct TestResult {
    bool success;
    string errorMessage;
}
```

**Comment by bclehmann** [comment]
> @bclehmann do you think this is worth adding to AvaPlot?

Yeah, I think it looks easy and helpful for a lot of users.

I don't really like how `GDI.DrawingTest` works, I think there's a better way to indicate a failure. Mostly from an API clarity perspective. And maybe issues like this must be handled by the caller under the threat of crashing due to an unhandled exception.

**Comment by swharden** [comment]
> I think there's a better way to indicate a failure.

Hey @bclehmann, I'm a bit confused by your objection/suggestion. Could you rephrase your concern or proposed solution to help me understand it better?

My intent is that `GDI.DrawingTest()` will only get called in designer mode, and only once when the control is first loaded. When a user drags a control from the toolbox and drops it onto the window, a System.Drawing support or version issue may cause it to throw an exception which Visual Studio has to catch and display (e.g., #993). I think those design-time exceptions are pretty jarring and confusing for new users (especially if the program builds and runs fine when the user presses F5).

I could move `DrawingTest()` somewhere in the `Control` namespace since it's not supposed to be used for any other purpose.

Interested in your feedback!
Scott

**Comment by bclehmann** [comment]
>  Could you rephrase your concern or proposed solution to help me understand it better?

I had two concerns, the first was the throw on [line 39](https://github.com/ScottPlot/ScottPlot/blob/f1d9e0851654507b5520680893aaf449cb501cdd/src/ScottPlot/Drawing/GDI.cs#L39) which looked like it was using try/catch as a control structure, but the catch can actually catch multiple issues so my conclusion was too hasty. One could argue about it as a matter of taste, but now that I have thought about it I think it's probably the better solution.

The other was how the function indicates the design-time exception. I think returning a string with an error message is counter-intuitive. The name `DrawingTest` implies it either passes or fails a test, which is generally shown through returning a boolean indicating whether it passed, or through throwing an exception in the event that it fails. Should the error message be necessary you can read it when you catch the exception. Also, normally returning `null` implies an error.

My preferred solution is just to change the return type to void and remove the catch statement. The caller catches any exceptions and displays the error message. However, I think returning a boolean (or a struct with a boolean success field) is more appropriate for a method with the word "Test" in its name.

> I could move DrawingTest() somewhere in the Control namespace since it's not supposed to be used for any other purpose.

I like this idea, but I think there is a benefit to keeping it in GDI, since moving to a different graphics library would require this to be rewritten. I like the current state where nothing (that I know of) in the Control namespace is dependent on System.Drawing.Common.

**Comment by bclehmann** [comment]
You can see what I meant in #1052. I don't know how well I communicated my concerns.

**Comment by swharden** [comment]
> returning a string with an error message is counter-intuitive

I'm sold! A try/catch block in the initializer of each control is a lot more straightforward, and a better design.

Thanks for expanding on your thoughts/suggestions!

I'll work in #1052 to test the AvaPlot and also add similar try/catch blocks to the other controls 👍
